brave/src/main/java/brave/handler/SpanCollector.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * <h3>Relationship to Span lifecycle</h3>
+ * The pair of {@link #begin} and {@link #end} seems the same as the span lifecycle. In
+ * most cases it will be the same, but you cannot assume this.
+ *
+ * <p>A {@link TraceContext} could be recorded twice, for example, if a long operation
+ * began, called {@link Span#flush()} (recording 1) and later called {@link Span#finish()}
+ * (recording 2). A {@link TraceContext} could be abrupted by garbage collection resulting in a
+ * {@link Cause#ABANDON}. A user could even {@linkplain Cause#ABANDON abandon} a span without
+ * recording anything!
+ *
+ * <p>Collectors that process finished spans will need to look at the {link Cause} and {@link
+ * MutableSpan} collected. For example, {@link Cause#FINISH} is usually a good enough heuristic to
+ * find complete spans.
+ *
+ * <h3>Advanced Notes</h3>
+ * <p>It is important to do work quickly as callbacks are run on the same thread as application
+ * code. However, do not mutate {@link MutableSpan} between callbacks, as it is not thread safe.
+ *
+ * <p>The {@link TraceContext} and {@link MutableSpan} parameter from {@link #begin} will be
+ * the same reference for {@link #end}.
+ *
+ * <p>If caching the context or span parameters between callbacks, consider a {@link WeakReference}
+ * to avoid holding up garbage collection.
+ *
+ * <p>The {@link #begin} callback primarily supports tracking of children, or partitioning of
+ * data for backend that needs to see an entire {@linkplain TraceContext#localRootId() local root}.
+ */
+public interface SpanCollector {
+  /** What ended the data collection? */
+  enum Cause {
+    /**
+     * Called on {@link Span#abandon()}.
+     *
+     * <p>This is useful when counting children. Decrement your counter when this occurs as the
+     * span will not be reported.
+     *
+     * <p><em>Note:</em>Abandoned spans should be ignored as they aren't indicative of an error.
+     * Some instrumentation speculatively create a span for possible outcomes such as retry.
+     */
+    ABANDON,
+    /**
+     * Called on {@link Span#finish()} and is the simplest cause to reason with. When {@link
+     * MutableSpan#startTimestamp()} is present, you can assume with high confidence you have all
+     * recorded data for this span.
+     */
+    FINISH,
+    /**
+     * Called on {@link Span#flush()}.
+     *
+     * <p>Even though the span here is incomplete (missing {@link MutableSpan#finishTimestamp()},
+     * it is reported to the tracing system unless a {@link FinishedSpanHandler} returns false.
+     */
+    FLUSH,
+    /**
+     * Called when the trace context was garbage collected prior to completion.
+     *
+     * <p>Unlike {@link FinishedSpanHandler#supportsOrphans()}, this is called even if empty.
+     * Non-empty spans are reported to the tracing system unless a {@link FinishedSpanHandler}
+     * returns false.
+     *
+     * @see FinishedSpanHandler#supportsOrphans()
+     */
+    ORPHAN
```

Comment:
the other constants are named linear to their `Span` methods, why `ORPHAN` for gc collected, are there other use cases leading to `ORPHAN` besides gc ?

Reply:
well GC is the side affect of orphaning your span (ex by creating it and dropping a reference to it). I think that's why we named it orphan back then, but I can look up notes if you like.

Main thing is the fact that it is GC'ed is only because someone orphaned it. If they held a reference it would not become GC'ed. Possibly I can clarify the description somewhere.